File: Java/chapter_10_challenge/app/src/main/java/com/glella/criminalintent/CrimeLab.java
package com.glella.criminalintent;

import android.content.Context;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class CrimeLab {

    private static CrimeLab sCrimeLab;

    // Challenge 2 - Using a Linked Hashmap we will be able to fetch a single crime with its UUID
    // Using a Linked HashMap so we can preserve the order of the Crimes
    //private List<Crime> mCrimes;
    private Map<UUID, Crime> mCrimes;


    public static CrimeLab get(Context context) {
        if (sCrimeLab == null) {
            sCrimeLab = new CrimeLab(context);
        }
        return sCrimeLab;
    }

    private CrimeLab(Context context) {
        // Challenge
        //mCrimes = new ArrayList<Crime>();
        mCrimes = new LinkedHashMap<>();

        for (int i = 0; i < 100; i++) {
            Crime crime = new Crime();
            crime.setTitle("Crime #" + i);
            crime.setSolved(i % 2 == 0); // Every other one
            // Challenge
            //mCrimes.add(crime);
            mCrimes.put(crime.getId(), crime);
        }
    }

    public List<Crime> getCrimes() {
        // Challenge
        //return mCrimes;
        return new ArrayList<>(mCrimes.values());
    }

    public Crime getCrime(UUID id) {
        // Challenge
        //for (Crime crime : mCrimes) {
            //if (crime.getId().equals(id)) {
                //return crime;
            //}
        //}
        //return null;

        return mCrimes.get(id);
    }

}

File: Kotlin/chapter_10_challenge/app/src/main/java/com/glella/criminalintent/CrimeLab.kt
package com.glella.criminalintent

import java.util.*
import kotlin.collections.LinkedHashMap

object CrimeLab {

    // Challenge 2 - Using a Linked Hashmap we will be able to fetch a single crime with its UUID
    // Using a Linked hashMap so we can preserve the order of the Crimes
    //var mCrimes = ArrayList<Crime>()
    var mCrimes = LinkedHashMap<UUID, Crime>()


    init {
        for (i in 0..99) {
            val crime = Crime()
            crime.mTitle = "Crime #$i"
            crime.mSolved = (i % 2 == 0) // Every other one
            // Challenge 2
            //mCrimes.add(crime)
            mCrimes.put(crime.mId, crime)
        }
    }

    // Challenge
    fun getCrimeList(): List<Crime> {
        return ArrayList(mCrimes.values)
    }

    fun getCrime(id: UUID): Crime? {
        // Challenge
        //for (item in mCrimes)
            //if (item.mId.equals(id)) return item
        //return null

        return mCrimes.get(id)
    }

}